### PR TITLE
#25326: ttnn.gelu accuracy should be improved

### LINKED ISSUE #25326: `ttnn.gelu` accuracy should be improved
### Describe the bug

`ttnn.gelu` seems to have accuracy problems in the range [-5, -2].
This seems to be a known problem( #21313  #23669 https://github.com/tenstorrent/tt-metal/issues/20562 ) but I want to have a separate tracking issue on Forge side.
This is a hard blocker for 3 variants of ViT (base patch32 384, large patch16 224 and large patch32 384) and Roberta trough tt-xla. It's interesting that some variants of ViT work fine with current levels of inaccuracy while others don't.
In apparent contradiction to investigations done in #21313 , when we use Forge tooling to bypass gelu and do it on the CPU, ViT passes.

### Steps to reproduce the issue

1. Run the following code
```
import torch
import ttnn
import numpy as np

def spacing(x: torch.Tensor):
    """Calculate spacing (ULP size) for bfloat16 and float32 inputs."""
    # If input is bfloat16, use abs(x) * 2^-7 for normal numbers
    if x.dtype == torch.bfloat16:
        ulp = torch.abs(x).to(torch.float32) * 2**-7
        # For zero, use smallest positive normal bfloat16 (2^-126)
        ulp = torch.where(x == 0, torch.tensor(2**-126, dtype=torch.float32), ulp)
        return ulp
    else:
        # For float32, use nextafter
        direction = torch.where(x >= 0, torch.inf, -torch.inf)
        next_val = torch.nextafter(x, direction)
        return torch.abs(next_val - x)

def ulp_difference(a, b):
    """Calculate ULP (Units in the Last Place) difference between two tensors for bfloat16 and float32."""
    # Ensure inputs are torch tensors
    if not torch.is_tensor(a):
        a = torch.tensor(a)
    if not torch.is_tensor(b):
        b = torch.tensor(b)
    # Move to CPU and ensure same device/dtype
    a = a.detach().cpu()
    b = b.detach().cpu()
    # Handle special cases (inf, nan)
    mask_valid = torch.isfinite(a) & torch.isfinite(b)
    if not torch.any(mask_valid):
        return torch.tensor([float('inf')])
    # Use the larger absolute value for spacing calculation
    max_abs = torch.maximum(torch.abs(a), torch.abs(b))
    # Calculate spacing (distance to next representable float)
    ulp_spacing = spacing(max_abs)
    # Calculate ULP difference
    abs_diff = torch.abs(a - b)
    ulp_diff = abs_diff / ulp_spacing
    return ulp_diff[mask_valid]

RANGE = 4

def test_ttnn_gelu():
    x_values = torch.linspace(-RANGE, RANGE, 2048, dtype=torch.bfloat16).reshape(-1, 32, 32)
    
    with ttnn.manage_device(0) as dev:
        # Convert to ttnn tensor with BFLOAT16 and tile layout
        ttnn_input = ttnn.from_torch(
            x_values,
            dtype=ttnn.bfloat16,
            layout=ttnn.TILE_LAYOUT,
            device=dev
        )
    
        ttnn_result = ttnn.gelu(ttnn_input)
        ttnn_output = ttnn.to_torch(ttnn_result)
        torch_reference = torch.nn.functional.gelu(x_values)
        
        print("\n" + "="*60)
        print("COMPARISON RESULTS")
        print("="*60)
        
        # Calculate various error metrics
        abs_diff = torch.abs(ttnn_output - torch_reference)
        rel_diff = abs_diff.float() / (torch.abs(torch_reference.float()) + 1e-9)
        
        print(f"Max absolute difference: {abs_diff.max():.6e}")
        print(f"Mean absolute difference: {abs_diff.mean():.6e}")
        print(f"Max relative difference: {rel_diff.max():.6e}")
        print(f"Mean relative difference: {rel_diff.mean():.6e}")
        
        # Calculate ULP difference
        ulp_diffs = ulp_difference(ttnn_output, torch_reference)
        
        print(f"\nULP Statistics:")
        print(f"Max ULP difference: {torch.max(ulp_diffs):.2f}")
        print(f"Mean ULP difference: {torch.mean(ulp_diffs):.2f}")
        print(f"Median ULP difference: {torch.median(ulp_diffs):.2f}")
        
        N=30
        # Print some sample comparisons (equally spaced subset)
        total_values = len(x_values.flatten())
        indices = np.linspace(0, total_values - 1, N, dtype=int)
        
        print(f"\nSample comparisons ({N} equally spaced values):")
        print("Input\t\tTorch gelu\tTTNN gelu\tAbs Diff\tULP Diff\tOne ULP size")
        print("-" * 70)
        for i in indices:
            inp = x_values.flatten()[i].item()
            torch_val = torch_reference.flatten()[i].item()
            ttnn_val = ttnn_output.flatten()[i].item()
            abs_d = abs_diff.flatten()[i].item()
            ulp_d = ulp_diffs[i] if i < len(ulp_diffs) else 0
            torch_ulp = spacing(torch.tensor(torch_val, dtype=torch.bfloat16)).item()
            print(f"{inp:8.3f}\t{torch_val:10.6f}\t{ttnn_val:10.6f}\t{abs_d:.2e}\t{ulp_d:8.2f}\t{torch_ulp:.2e}")
        
        reasonable_ulp = torch.sum(ulp_diffs < 50) / len(ulp_diffs) * 100
        print(f"\nPercentage of values with ULP < 50: {reasonable_ulp:.1f}%")


if __name__ == "__main__":
    test_ttnn_gelu()
```

### Expected behavior

I don't know what level of accuracy I need, but I need it improved. Attached code is for illustration purposes only.

### Please complete the following environment information

Ubuntu 22.04
Metal hash d5c259dd7bc8d0c9879a4389d50e65c321cc0c6f

### PR DESCRIPTION
**Pull request by mcw-vishal**
### Ticket
#25326 

### Problem description
ttnn.gelu default implementation needs accuracy improvement.

### What's changed
Proposed a new checbychev polyval15 approach to improve accuracy.

# Chebyshev Implementation Performance Analysis

## Overview
Performance comparison of new Chebyshev approximation approaches (BH, WH) against baseline implementations.

## Baseline vs Chebyshev Approaches

| Implementation | 1 Tile | 1K Tensor | 8K Tensor | Type |
|----------------|--------|-----------|-----------|------|
| **Main Branch (WH)** | 4,465 ns | 52,048 ns | 3,054,703 ns | 🔵 Baseline |
| **Main Branch Approx (WH)** | 1,866 ns | 26,849 ns | 1,340,607 ns | 🔵 Baseline |
| **Chebyshev BH** | 2,654 ns | 21,584 ns | 844,148 ns | 🟢 New Approach |
| **Chebyshev WH** | 3,841 ns | 41,749 ns | 2,373,080 ns | 🟢 New Approach |

### vs Main Branch  Baseline
| Chebyshev Approach | 1 Tile | 1K Tensor | 8K Tensor |
|-------------------|--------|-----------|-----------|
| **BH Implementation** | 🟢 **1.68x faster** | 🟢 **2.41x faster** | 🟢 **3.62x faster** |
| **WH Implementation** | 🟢 **1.16x faster** | 🟢 **1.25x faster** | 🟢 **1.29x faster** |

### vs Main Branch Approx Baseline
| Chebyshev Approach | 1 Tile | 1K Tensor | 8K Tensor |
|-------------------|--------|-----------|-----------|
| **BH Implementation** | 🔴 **1.42x slower** | 🟢 **1.24x faster** | 🟢 **1.59x faster** |
| **WH Implementation** | 🔴 **2.06x slower** | 🔴 **1.56x slower** | 🔴 **1.77x slower** |

For detailed info kindly refer #20562 

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17605554877) CI passes
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/17605535611l) CI with demo tests passes (if applicable)
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/runs/17756362840) one test failed same as [main](https://github.com/tenstorrent/tt-metal/actions/runs/17752478737) 
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/runs/17756373253) CI passes (if applicable)
- [ ] [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/17756782982/job/50461546082).
- [x] [Single Card Frequent tests](https://github.com/tenstorrent/tt-metal/actions/runs/17761529260)
- [x] [T3000 Pipeline](https://github.com/tenstorrent/tt-metal/actions/runs/17605508237) T3K demo failures confirmed with model owners are unrelated to this PR changes, T3K frequent failed test also failed in [main](https://github.com/tenstorrent/tt-metal/actions/runs/17619630009).
- [x] [T3000 Model perf](https://github.com/tenstorrent/tt-metal/actions/runs/17680604721)

### CODE REVIEW

**Review by KalaivaniMCW — APPROVED** [review]

**Comment by pavlejosipovic** [comment]
Pls post passing pipelines for these
https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml
https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml
https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml

**Comment by mcw-vishal** [comment]
> Pls post passing pipelines for these https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml https://github.com/tenstorrent/tt-metal/actions/workflows/fast-dispatch-full-regressions-and-models.yaml

Hi @pavlejosipovic, I have updated the CI runs for perf-device and perf-models. Triggered the frequent model CI, will update once finished.

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h [hunk 1]
```diff
@@ -22,9 +72,21 @@ void gelu_derivative_init() {
 
 template <bool APPROXIMATION_MODE, int ITERATIONS = 8>
```
Comment (mcw-anasuya) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h:73]: We could follow llk ckernels and not hardcode 8 here. Instead pass it as given [here](https://github.com/tenstorrent/tt-metal/pull/27829/files#diff-ccdd2c72881127d334359db2ccfc678574cebc89bce063a745f0e6d12eb636a4R24)

**Review by mcw-anasuya — APPROVED** [review]

**Review by VirdhatchaniKN — APPROVED** [review]

File: tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h [hunk 2]
```diff
@@ -10,6 +10,56 @@
...
+        (c4)) *                                                                                                    \
+           (x) +                                                                                                   \
+       (c3)) *                                                                                                     \
+          (x) +                                                                                                    \
+      (c2)) *                                                                                                      \
+         (x) +                                                                                                     \
+     (c1)) * (x) +                                                                                                 \
+        (c0)
+
+inline sfpi::vFloat calculate_gelu_chebychev(sfpi::vFloat val) {
```
Comment (yieldthought) [inline tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h]: chebyshev with an s

**Review by nvelickovicTT — APPROVED** [review]

**Review by mbahnasTT — APPROVED** [review]

**Review by pavlejosipovic — APPROVED** [review]

**Review by gwangTT — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h [hunk 3]
```diff
@@ -10,6 +10,56 @@
...
+         (c5)) *                                                                                                   \
+            (x) +                                                                                                  \
+        (c4)) *                                                                                                    \
+           (x) +                                                                                                   \
+       (c3)) *                                                                                                     \
+          (x) +                                                                                                    \
+      (c2)) *                                                                                                      \
+         (x) +                                                                                                     \
+     (c1)) * (x) +                                                                                                 \
+        (c0)
```
Comment (patrickroberts) [inline tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h:31]: Could you maybe extract this to a common header? Also suggestion for simplifying maybe?

```cpp
template <typename V, typename... Ts>
inline V polyval(V x, Ts... coefs) {
  static_assert(sizeof...(Ts) >= 2, "Must be at least two coefficients");
  constexpr auto recur = [](auto self, V val, V acc, auto coef, auto... rest) -> V {
    if constexpr (sizeof...(rest) == 0) {
      return acc * val + coef;
    } else {
      return self(self, val, acc * val + coef, rest...);
    }
  };

  // coefs... unpacks as acc, coef, rest...
  return recur(recur, x, coefs...);
}

// usage
result = polyval(
    val,
    -1.81205228163e-09,
    -4.59055119276e-08,
    -3.74540617693e-07,
    -2.29754133825e-07,
    1.19076782913e-05,
    4.25116466215e-05,
    -0.000138391838381,
    -0.000862052441087,
    0.000768340223025,
    0.0092074331601,
    -0.00208478037614,
    -0.0656369476513,
    0.00244542739174,
    0.398579460781,
    0.499174645395,
    2.98325768482e-05);
```

**Review by dchenTT — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h (hunk 3 above)
Comment (patrickroberts) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h:31]: After looking at Compiler Explorer, there are a couple adjustments that should be made to my suggestion to ensure it has same performance as the macro: https://godbolt.org/z/v9sMs5qE4
- The coefficients must be passed as `float`, not `double`
- All parameters must be accepted as `const T &` so that `inline` actually inlines the function
- `val` and `acc` must be passed as `const V &` so that the code gets better optimized

Putting that together, the function now looks like this:
```cpp
template <typename V, typename... Ts>
inline V polyval(const V& x, const Ts&... coefs) {
    static_assert((... and std::is_same_v<Ts, float>),
                  "Must not pass coefficients as double");
    static_assert(sizeof...(Ts) >= 2, "Must pass at least two coefficients");
    constexpr auto recur = [](auto self, const V& val, const V& acc, auto coef,
                              auto... rest) -> V {
        if constexpr (sizeof...(rest) == 0) {
            return acc * val + coef;
        } else {
            return self(self, val, acc * val + coef, rest...);
        }
    };

    // coefs... unpacks as acc, coef, rest...
    return recur(recur, x, coefs...);
}
```

**Review by yieldthought — APPROVED** [review]

File: tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h (hunk 3 above)
Comment (mcw-vishal) [reply tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_sfpu/ckernel_sfpu_gelu.h:31]: @patrickroberts I am testing this on a separate branch, since I would have to record perf with change in implementation. Any perf change would affect the model metric to be updated. Once I have the new metric I will update these changes in a separate PR. Thanks.

**Comment by mcw-vishal** [comment]
There were some changes pushed in main ,due to which llama-N300-perf (prefill_time_to_token) and qwen25vl N300 tests (expected text needs to be modified) are failing (other failures are unrelated). Since this is a priority ticket I am going to merge with the main branch. I will update the corresponding targets in a follow-up PR. 

Single Card Demo [With perf](https://github.com/tenstorrent/tt-metal/actions/runs/17819497842/job/50659779241) 
Latest APC Link [passed](https://github.com/tenstorrent/tt-metal/actions/runs/17821472755)

**Comment by mcw-vishal** [comment]
> There were some changes pushed in main ,due to which llama-N300-perf (prefill_time_to_token) and qwen25vl N300 tests (expected text needs to be modified) are failing (other failures are unrelated). Since this is a priority ticket I am going to merge with the main branch. I will update the corresponding targets in a follow-up PR.
> 
> Single Card Demo [With perf](https://github.com/tenstorrent/tt-metal/actions/runs/17819497842/job/50659779241) Latest APC Link [passed](https://github.com/tenstorrent/tt-metal/actions/runs/17821472755)

Update: BH Post commit [PASSED](https://github.com/tenstorrent/tt-metal/actions/runs/17825918007). Going ahead with the merge as mentioned in above comment.
